eclipse-theia/theia
# Refactored LanguagesMainImpl and OutputChannelRegistryMainImpl so they could be extended

#### What it does

This PR introduces LanguagesMainFactory and an OutputChannelRegistryFactory so that we can extend LanguagesMainImpl and OutputChannelRegistryMainImpl in order to provide metrics for the plugin extension.

See: [this](https://spectrum.chat/theia/dev/current-state-of-metrics-in-theia~8efd05d2-a25b-4629-a1b1-1dd8a0ccef96?m=MTU2NzE0MDIxMjMwMg==) for the implementation way and [this](https://spectrum.chat/theia/dev/current-state-of-metrics-in-theia~8efd05d2-a25b-4629-a1b1-1dd8a0ccef96) for the whole discussion.

#### How to test

This PR doesn't include any new functionality, more just a restructuring of everything that enables it to be extended more.

With that said, you can test by installing a vscode extension that has lots of features such as vscode-java and testing whether or not the language features work on a simple project such as: https://github.com/che-samples/console-java-simple.

For reference, I tested auto completion, validation, find all references, go to definition, go to symbol, go to type definition, format document, refactor, rename symbol and everything was working as expected.

Additionally, if you want to see how it's extended you can build and use this theia extension: https://github.com/JPinkney/theia-plugin-extension-metrics

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
